File: openwire-cpp/src/main/cpp/activemq/Connection.cpp
/*
 * Copyright 2006 The Apache Software Foundation or its licensors, as
 * applicable.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "activemq/Connection.hpp"
#include "activemq/Session.hpp"

using namespace apache::activemq;


// --- Constructors -------------------------------------------------

/*
 *
 */
Connection::Connection(p<ITransport> transport, p<ConnectionInfo> connectionInfo)
{
    this->transport               = transport ;
    this->connectionInfo          = connectionInfo ;
    this->acknowledgementMode     = AutoAckMode ;
    this->connected               = false ;
    this->closed                  = false ;
    this->brokerInfo              = NULL ;
    this->brokerWireFormat        = NULL ;
    this->sessionCounter          = 0 ;
    this->tempDestinationCounter  = 0 ;
    this->localTransactionCounter = 0 ;

    // Hook up as a command listener and start dispatching
    transport->setCommandListener(smartify(this)) ;
    transport->start() ;
}

/*
 *
 */
Connection::~Connection()
{
    // Make sure connection is closed
    close() ;
}


// --- Attribute methods --------------------------------------------

/*
 *
 */
void Connection::setExceptionListener(p<IExceptionListener> listener)
{
    this->listener = listener ;
}

/*
 *
 */
p<IExceptionListener> Connection::getExceptionListener()
{
    return listener ;
}

/*
 *
 */
p<ITransport> Connection::getTransport()
{
    return transport ;
}

/*
 *
 */
void Connection::setTransport(p<ITransport> transport)
{
    this->transport = transport ;
}

/*
 *
 */
p<string> Connection::getClientId()
{
    return connectionInfo->getClientId() ;
}

/*
 *
 */
void Connection::setClientId(const char* value) throw (CmsException)
{
    if( connected )
        throw CmsException("You cannot change the client id once a connection is established") ; 

    p<string> clientId = new string(value) ;
    connectionInfo->setClientId( clientId ) ;
}


/*
 *
 */
p<BrokerInfo> Connection::getBrokerInfo()
{
    return brokerInfo ;
}

/*
 *
 */
p<WireFormatInfo> Connection::getBrokerWireFormat()
{
    return brokerWireFormat ;
}

/*
 *
 */
AcknowledgementMode Connection::getAcknowledgementMode()
{
    return acknowledgementMode ;
}

/*
 *
 */
void Connection::setAcknowledgementMode(AcknowledgementMode ackMode)
{
    acknowledgementMode = ackMode ;
}

/*
 *
 */
p<ConnectionId> Connection::getConnectionId()
{
    return connectionInfo->getConnectionId() ;
}


// --- Operation methods --------------------------------------------

/*
 *
 */
void Connection::close()
{
    if( !closed )
    {
        list< p<ISession> >::iterator sessionIter ;

        // Iterate through all sessions and close them down
        for( sessionIter = sessions.begin() ;
             sessionIter != sessions.end() ;
             sessionIter++ )
        {
            (*sessionIter)->close() ;
        }
        // Empty session list
        sessions.clear() ;

        // Remove connection from broker
        disposeOf( getConnectionId() ) ;

        // Finally, transmit a shutdown command to broker
        transport->oneway( new ShutdownInfo() ) ;
        closed = true ;
    }
}

/*
 *
 */
p<ISession> Connection::createSession() throw(CmsException)
{
    return createSession(acknowledgementMode) ;
}

/*
 *
 */
p<ISession> Connection::createSession(AcknowledgementMode ackMode) throw(CmsException)
{
    p<SessionInfo> sessionInfo = createSessionInfo( ackMode ) ;

    // Send session info to broker
    syncRequest(sessionInfo) ;

    p<ISession> session = new Session(smartify(this), sessionInfo, ackMode) ;
    sessions.push_back(session) ;

    return session ;
}

/*
 * Performs a synchronous request-response with the broker.
 */
p<Response> Connection::syncRequest(p<ICommand> command) throw(CmsException) 
{
    checkConnected() ;
    
    p<Response> response = transport->request(command) ;

    if( response->getDataStructureType() == ExceptionResponse::TYPE )
    {
        p<ExceptionResponse> exceptionResponse = p_cast<ExceptionResponse> (response) ;
        p<BrokerError>       brokerError = exceptionResponse->getException() ;
        string               message ;

        // Build error message
        message.assign("Request failed: ") ;
        message.append( brokerError->getExceptionClass()->c_str() ) ;
        message.append(", ") ;
        message.append( brokerError->getStackTrace()->c_str() ) ;

        // TODO: Change to CMSException()
        throw CmsException( message.c_str() ) ; 
    }
    return response ;
}

/*
 *
 */
void Connection::oneway(p<ICommand> command) throw(CmsException)
{
    checkConnected() ;
    transport->oneway(command) ;
}

/*
 *
 */
void Connection::disposeOf(p<IDataStructure> dataStructure) throw(CmsException)
{
    p<RemoveInfo> command = new RemoveInfo() ;
    command->setObjectId( dataStructure ) ;
    syncRequest(command) ;
}

/*
 * Creates a new temporary destination name.
 */
p<string> Connection::createTemporaryDestinationName()
{
    p<string> name = new string() ;
    char*     buffer = new char[15] ;

    {
        LOCKED_SCOPE( mutex );

        name->assign( connectionInfo->getConnectionId()->getValue()->c_str() ) ;
        name->append( ":" ) ;
#ifdef unix
        sprintf(buffer, "%lld", ++tempDestinationCounter) ;
#else
        sprintf(buffer, "%I64d", ++tempDestinationCounter) ;
#endif
        name->append( buffer ) ;
    }

    return name ;
}

/*
 * Creates a new local transaction ID.
 */
p<LocalTransactionId> Connection::createLocalTransactionId()
{
    p<LocalTransactionId> id = new LocalTransactionId() ;

    id->setConnectionId( getConnectionId() ) ;

    {
        LOCKED_SCOPE (mutex);
        id->setValue( ++localTransactionCounter ) ;
    }

    return id ;
}


// --- Implementation methods ---------------------------------------

/*
 *
 */
p<SessionInfo> Connection::createSessionInfo(AcknowledgementMode ackMode)
{
    p<SessionInfo> sessionInfo = new SessionInfo() ;
    p<SessionId>   sessionId   = new SessionId() ;
    
    sessionId->setConnectionId ( connectionInfo->getConnectionId()->getValue() ) ;

    {
        LOCKED_SCOPE( mutex );
        sessionId->setValue( ++sessionCounter ) ; 
    }

    sessionInfo->setSessionId( sessionId ) ;
    return sessionInfo ; 
}

/*
 *
 */
void Connection::checkConnected() throw(CmsException) 
{
    if( closed )
        throw ConnectionClosedException("Oops! Connection already closed.") ;

    if( !connected )
    {
        connected = true ;

        // Send the connection info and see if we get an ack/nak
        syncRequest( connectionInfo ) ;
    } 
}

/*
 * Handle incoming commands.
 */
void Connection::onCommand(p<ITransport> transport, p<ICommand> command)
{
    if( command->getDataStructureType() == MessageDispatch::TYPE )
    {
        p<MessageDispatch> dispatch = p_cast<MessageDispatch> (command) ;
        p<ConsumerId> consumerId = dispatch->getConsumerId() ;
        p<MessageConsumer> consumer = NULL ;
        list< p<ISession> >::const_iterator tempIter ;

        // Iterate through all sessions and check if one has the consumer
        for( tempIter = sessions.begin() ;
             tempIter != sessions.end() ;
             tempIter++ )
        {
            consumer = p_cast<Session> (*tempIter)->getConsumer(consumerId) ;

            // Found a match?
            if( consumer != NULL )
                break ;
        }
        if( consumer == NULL )
            cout << "ERROR: No such consumer active: " << consumerId->getValue() << endl ;
        else
        {
            p<ActiveMQMessage> message = p_cast<ActiveMQMessage> (dispatch->getMessage()) ;
            consumer->dispatch(message) ;
        }
    }
    else if( command->getDataStructureType() == WireFormatInfo::TYPE )
        this->brokerWireFormat = p_cast<WireFormatInfo> (command) ;

    else if( command->getDataStructureType() == BrokerInfo::TYPE )
        this->brokerInfo = p_cast<BrokerInfo> (command) ;

    else
        cout << "ERROR: Unknown command: " << command->getDataStructureType() << endl ;
}

/*
 * Handle incoming broker errors.
 */
void Connection::onError(p<ITransport> transport, exception& error)
{
    if( listener != NULL )
        this->listener->onException(error) ;
    else
        cout << "ERROR: Received a broker exception: " << error.what() << endl ;
}

File: activemq-dotnet/src/main/csharp/ActiveMQ/Connection.cs
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
using ActiveMQ.Commands;
using ActiveMQ.Transport;
using NMS;
using System;
using System.Collections;

namespace ActiveMQ
{
    /// <summary>
    /// Represents a connection with a message broker
    /// </summary>
    public class Connection : IConnection
    {
        private ITransport transport;
        private ConnectionInfo info;
        private AcknowledgementMode acknowledgementMode = AcknowledgementMode.AutoAcknowledge;
        private BrokerInfo brokerInfo; // from broker
        private WireFormatInfo brokerWireFormatInfo; // from broker
        private IList sessions = new ArrayList();
        private IDictionary consumers = new Hashtable(); // TODO threadsafe
        private bool connected;
        private bool closed;
        private long sessionCounter;
        private long temporaryDestinationCounter;
        private long localTransactionCounter;
        
        
        public Connection(ITransport transport, ConnectionInfo info)
        {
            this.transport = transport;
            this.info = info;
            this.transport.Command = new CommandHandler(OnCommand);
			this.transport.Exception = new ExceptionHandler(OnException);
            this.transport.Start();
        }
        
        /// <summary>
        /// Starts message delivery for this connection.
        /// </summary>
        public void Start()
        {
        }
        
        
        /// <summary>
        /// Stop message delivery for this connection.
        /// </summary>
        public void Stop()
        {
        }
        
        /// <summary>
        /// Creates a new session to work on this connection
        /// </summary>
        public ISession CreateSession()
        {
            return CreateSession(acknowledgementMode);
        }
        
        /// <summary>
        /// Creates a new session to work on this connection
        /// </summary>
        public ISession CreateSession(AcknowledgementMode acknowledgementMode)
        {
            SessionInfo info = CreateSessionInfo(acknowledgementMode);
            SyncRequest(info);
            Session session = new Session(this, info, acknowledgementMode);
            sessions.Add(session);
            return session;
        }
        
        public void Dispose()
        {
            /*
            foreach (Session session in sessions)
            {
                session.Dispose();
            }
            */
            DisposeOf(ConnectionId);
            sessions.Clear();
			transport.Oneway(new ShutdownInfo());
            transport.Dispose();
            closed = true;
        }
        
        // Properties
        
        public ITransport ITransport
        {
            get { return transport; }
            set { this.transport = value; }
        }
        
        
        public AcknowledgementMode AcknowledgementMode
        {
            get { return acknowledgementMode; }
            set { this.acknowledgementMode = value; }
        }
        
        public string ClientId
        {
            get { return info.ClientId; }
            set {
                if (connected)
                {
                    throw new NMSException("You cannot change the ClientId once the Connection is connected");
                }
                info.ClientId = value;
            }
        }
        
        public ConnectionId ConnectionId
        {
            get {
                return info.ConnectionId;
            }
        }
        
        public BrokerInfo BrokerInfo
        {
            get {
                return brokerInfo;
            }
        }
        
        public WireFormatInfo BrokerWireFormat
        {
            get {
                return brokerWireFormatInfo;
            }
        }
        
        // Implementation methods
        
        /// <summary>
        /// Performs a synchronous request-response with the broker
        /// </summary>
        public Response SyncRequest(Command command)
        {
            CheckConnected();
            Response response = transport.Request(command);
            if (response is ExceptionResponse)
            {
                ExceptionResponse exceptionResponse = (ExceptionResponse) response;
                // TODO include stack trace
                throw new NMSException("Request failed: " + exceptionResponse);
            }
            return response;
        }
        
        public void OneWay(Command command)
        {
            CheckConnected();
            transport.Oneway(command);
        }
        
        public void DisposeOf(DataStructure objectId)
        {
            RemoveInfo command = new RemoveInfo();
            command.ObjectId = objectId;
            SyncRequest(command);
        }
        
        
        /// <summary>
        /// Creates a new temporary destination name
        /// </summary>
        public String CreateTemporaryDestinationName()
        {
            lock (this)
            {
                return info.ConnectionId.Value + ":" + (++temporaryDestinationCounter);
            }
        }
        
        /// <summary>
        /// Creates a new local transaction ID
        /// </summary>
        public LocalTransactionId CreateLocalTransactionId()
        {
            LocalTransactionId id= new LocalTransactionId();
            id.ConnectionId = ConnectionId;
            lock (this)
            {
                id.Value = (++localTransactionCounter);
            }
            return id;
        }
        
        protected void CheckConnected()
        {
            if (closed)
            {
                throw new ConnectionClosedException();
            }
            if (!connected)
            {
                connected = true;
                // now lets send the connection and see if we get an ack/nak
                SyncRequest(info);
            }
        }
        
        /// <summary>
        /// Register a new consumer
        /// </summary>
        /// <param name="consumerId">A  ConsumerId</param>
        /// <param name="consumer">A  MessageConsumer</param>
        public void AddConsumer(ConsumerId consumerId, MessageConsumer consumer)
        {
            consumers[consumerId] = consumer;
        }
        
        
        /// <summary>
        /// Remove a consumer
        /// </summary>
        /// <param name="consumerId">A  ConsumerId</param>
        public void RemoveConsumer(ConsumerId consumerId)
        {
            consumers[consumerId] = null;
        }
        
        
        /// <summary>
        /// Handle incoming commands
        /// </summary>
        /// <param name="transport">An ITransport</param>
        /// <param name="command">A  Command</param>
        protected void OnCommand(ITransport transport, Command command)
        {
            if (command is MessageDispatch)
            {
                MessageDispatch dispatch = (MessageDispatch) command;
                ConsumerId consumerId = dispatch.ConsumerId;
                MessageConsumer consumer = (MessageConsumer) consumers[consumerId];
                if (consumer == null)
                {
                    Console.WriteLine("ERROR: No such consumer active: " + consumerId);
                }
                else
                {
                    ActiveMQMessage message = (ActiveMQMessage) dispatch.Message;
                    consumer.Dispatch(message);
                }
            }
            else if (command is WireFormatInfo)
            {
                this.brokerWireFormatInfo = (WireFormatInfo) command;
            }
            else if (command is BrokerInfo)
            {
                this.brokerInfo = (BrokerInfo) command;
            }
            else
            {
                Console.WriteLine("ERROR: Unknown command: " + command);
            }
        }
		
		protected void OnException(ITransport sender, Exception exception) {
			Console.WriteLine("ERROR: Transport Exception: " + exception);
		}
		
        
        protected SessionInfo CreateSessionInfo(AcknowledgementMode acknowledgementMode)
        {
            SessionInfo answer = new SessionInfo();
            SessionId sessionId = new SessionId();
            sessionId.ConnectionId = info.ConnectionId.Value;
            lock (this)
            {
                sessionId.Value = ++sessionCounter;
            }
            answer.SessionId = sessionId;
            return answer;
        }
        
    }
}
